monai/metrics/cumulative_average.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2020 - 2021 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+
+from monai.metrics.utils import do_metric_reduction
+from monai.utils import MetricReduction
+
+from .metric import Cumulative
+
+
+class CumulativeAverage(Cumulative):
+    """
+    Cumulatively record data value and aggregate for the average value.
+    It supports single class or multi-class data, for example,
+    value can be 0.44 (like loss) or [0.3, 0.4] (like metrics of 2 classes).
+    It also supports distributed data parallel, sync data when aggregating.
+    For example, recording loss value and compute the oveall average value in every 5 iterations:
+
+    .. code-block:: python
+
+        average = CumulativeAverage()
+        for i, d in enumerate(dataloader):
+            loss = ...
+            average.append(loss)
+            if i % 5 == 0:
+                print(f"cumulative average of loss: {average.aggregate()}")
+        average.reset()
+
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.sum = None
+        self.not_nans = None
+
+    def reset(self):
+        """
+        Reset all the running status, including buffers, sum, not nans count, etc.
+
+        """
+        super().reset()
+        self.sum = None
+        self.not_nans = None
+
+    def aggregate(self):  # type: ignore
+        """
+        Sync data from all the ranks and compute the average value with previous sum value.
+
+        """
+        data = self.get_buffer()
+        if not isinstance(data, torch.Tensor):
+            raise ValueError("the data to aggregate must be PyTorch Tensor.")
+
+        # do metric reduction
+        f, not_nans = do_metric_reduction(data, reduction=MetricReduction.SUM_BATCH)
+        # clear the buffer for next update
+        super().reset()
```

Comment:
why reset the buffer here? in my opinion `self.sum` and `self.non_nans` could be local variables and they could always be computed from `self._buffers`?

Reply:
Hi @wyli ,

Actually, @myron and I also discussed about this logic today, it looks more clear to always compute from `self.buffers`, but it's better to clear the buffer after every aggregation and only keep the `sum` & `not_nans` for the next computation to **save memory** and **performance**, otherwise, we need to repeatedly compute the average value from all the buffer when aggregating.
What do you think?

Thanks.

Reply:
To compute the running average perhaps this class should override the append and extend?

Reply:
Hi @wyli ,

OK, that's an interesting idea, let me think about it.
The only obvious difference is that we need to sync across ranks before computation.

Thanks.

Reply:
@wyli I agree with Nic about clearing memory. We need to do aggregation for logging frequently (every iteration), if don't clear, then the GPU memory array will grow constantly and require sync across gpu on a larger size every time.
@Nic-Ma if it's easy to chage append (you can try), but I need it for MIL the sooner the better

Reply:
Hi @wyli ,

I updated the PR based on the online discussion, could you please help review it again?

Thanks in advance.